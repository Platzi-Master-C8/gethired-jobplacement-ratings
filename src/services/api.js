import config from '../config';
import list from '../data/listReviews';
import overallReviews from '../data/overallReviews';

const getData = (endpoint, options) => {
  const ops = {
    headers: config.headers,
    ...options
  };

  const url = `${config.api}/${endpoint}`;
  return fetch(url, ops)
};

const sendData = (endpoint, data) => {
  const url = `${config.api}/${endpoint}`;

  return fetch(url, {
    method: 'POST',
    body: JSON.stringify(data),
    headers: config.headers
  });
}

const api = {
  companyEvaluations: {
    listReviews(companyId, options = {}) {
      // return getData(`companyEvaluations/${companyId}`, options);
      return getData(`companyEvaluations`, options);
    },
    sendReview(companyId, data = {}) {
      return sendData(`companyEvaluations`, data);
    },
    mockDataList() {
      return list.map((review) => ({
        id: review.id,
        contentType: review.content_type,
        createdAt: review.created_at,
        isStillWorkingHere: review.is_still_working_here,
        jobTitle: review.job_title,
        nonUtilityCounter: review.non_utility_counter,
        utilityCounter: review.utility_counter,
        weightedAveragePerEvaluation: review.weighted_average_per_evaluation,
      }));
    },
    mockDataOverallReview() {
      return new Promise(resolve => resolve(...overallReviews));
    }
  },
};

export default api;
