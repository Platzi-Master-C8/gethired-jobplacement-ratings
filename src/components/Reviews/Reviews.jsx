import React, { Fragment, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { Box, Pagination, Stack } from '@mui/material';
import { ReviewCard } from '../ReviewCard';
import { FilterReviews } from '../FilterReviews';
import { SideInfo } from '../SideInfo';

import api from '../../services/api';
import config from '../../config';

const Reviews = ({ isMobile }) => {
    const [reasons, setReasons] = useState([]);
    const [error, setError] = useState(null);
    const [isLoaded, setIsLoaded] = useState(false);
    const [page, setPage] = useState(1);
    const [reviewsCount, setReviewsCount] = useState(0);
    const [list, setList] = useState([]);
    const [data, setData] = useState([]);
    const [sortCriteria, setSortCriteria] = useState({ sortKey: 'created_at', orientation: 'asc' });

    useEffect(() => {
        fetch(`${config.api}reporting-reason-types`)
            .then((res) => res.json())
            .then((obj) => setReasons(obj));
    }, []);

    useEffect(() => {
        api.companyEvaluations
            .listReviews(1, page)
            .then((response) => response.json())
            .then((result) => {
                setIsLoaded(true);
                setList(result.items);
                setData(result.items);
                setReviewsCount(result.total);
            })
            .catch((e) => {
                setIsLoaded(true);
                setError(e);
            });
    }, [page]);

    const handlePage = (e, v) => {
        setIsLoaded(false);
        setPage(v);
        window.scrollTo({ top: 0, behavior: 'smooth' });
    };

    const handleSearch = (query, attribute) => {
        setList(data.filter((x) => x[attribute].toLowerCase().includes(query.toLowerCase().trim())));
    };

    const toggleSortCriteria = (sortKey) => {
        const orientation =
            sortCriteria.sortKey === sortKey ? (sortCriteria.orientation === 'asc' ? 'desc' : 'asc') : 'asc';
        const sortedList = list.sort((a, b) =>
            a[sortKey] > b[sortKey] ? (orientation === 'asc' ? 1 : -1) : orientation === 'asc' ? -1 : 1,
        );
        setList(sortedList);
        setSortCriteria({ sortKey, orientation });
    };

    if (error) {
        return <div>Error: {error.message}</div>;
    }
    if (!isLoaded) {
        return <div>Loading...</div>;
    }

    return (
        <Fragment>
            <FilterReviews
                reviewsQuantity={list.length}
                reviewsCount={reviewsCount}
                handleSearch={handleSearch}
                sortCriteria={sortCriteria}
                toggleSortCriteria={toggleSortCriteria}
            />
            <Box
                sx={{
                    display: 'grid',
                    gap: 2,
                    gridTemplateColumns: 'repeat(auto-fill, minmax(min(100%, 25rem), 1fr))',
                    p: 1,
                    my: 2,
                }}
            >
                {list.map((review) => (
                    <ReviewCard key={review.id} review={review} reasons={reasons} />
                ))}
            </Box>
            <SideInfo isMobile={isMobile} />
            <Box
                sx={{
                    mb: 3,
                    display: 'grid',
                    justifyContent: 'center',
                }}
            >
                <Stack spacing={2}>
                    <Pagination
                        color="secondary"
                        count={Math.ceil(reviewsCount / 10)}
                        disabled={Math.ceil(reviewsCount / 10) < 1}
                        page={page}
                        onChange={handlePage}
                    />
                </Stack>
            </Box>
        </Fragment>
    );
};

Reviews.propTypes = {
    isMobile: PropTypes.bool.isRequired,
};

export default Reviews;
