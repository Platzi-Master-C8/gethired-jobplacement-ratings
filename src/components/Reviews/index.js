import Reviews from "./Reviews";

export { Reviews };
