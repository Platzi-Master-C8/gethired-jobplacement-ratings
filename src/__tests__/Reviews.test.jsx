import React from 'react';

import { render, waitFor } from '@testing-library/react';
import { Reviews } from 'Components/Reviews';
import api from '../services/api';

describe('Component <Reviews/>', () => {
    beforeEach(() => {
        global.fetch = jest.fn(() =>
            Promise.resolve({
                json: () => Promise.resolve(api.companyEvaluations.mockDataList()),
            }),
        );
    });

    test('renders Reviews', () => {
        const { asFragment } = render(<Reviews isMobile={false} />);
        waitFor(() => {
            expect(asFragment()).toMatchSnapshot();
        });
    });
});
