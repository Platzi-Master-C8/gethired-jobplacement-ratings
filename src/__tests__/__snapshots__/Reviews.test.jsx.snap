// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <Reviews/> renders Reviews 1`] = `
<DocumentFragment>
  <div>
    Loading...
  </div>
</DocumentFragment>
`;
